File: src/main/osu/parsers/osdb.ts
import type { OsdbData } from "@shared/types";
import { BinaryReader } from "./binary_reader";

import fs from "fs";
import zlib from "zlib";

const OSDB_FOOTER = "By Piotrekol";

const map_version_to_code = (version: string): number => {
    switch (version) {
        case "o!dm":
            return 1;
        case "o!dm2":
            return 2;
        case "o!dm3":
            return 3;
        case "o!dm4":
            return 4;
        case "o!dm5":
            return 5;
        case "o!dm6":
            return 6;
        case "o!dm7":
            return 7;
        case "o!dm8":
            return 8;
        case "o!dm7min":
            return 1007;
        case "o!dm8min":
            return 1008;
        default:
            return 1;
    }
};

const is_minimal_version = (version: string): boolean => {
    return version.endsWith("min");
};

const uses_gzip_payload = (version_code: number): boolean => {
    return version_code >= 7;
};

export class OsdbParser extends BinaryReader {
    private location: string = "";
    private data: OsdbData = {
        version_string: "o!dm8min",
        save_data: 0n,
        last_editor: "",
        count: 0,
        collections: []
    };

    parse = async (location: string): Promise<this> => {
        const raw = fs.readFileSync(location);
        this.location = location;
        this.set_buffer(raw);

        const version_string = this.plain_string();
        const version_code = map_version_to_code(version_string);

        if (uses_gzip_payload(version_code)) {
            const compressed = this.buffer.subarray(this.offset);
            const uncompressed = zlib.gunzipSync(compressed);
            this.set_buffer(uncompressed);
            this.plain_string();
        }

        const is_minimal = is_minimal_version(version_string);

        const save_data = this.long();
        const last_editor = this.plain_string();
        const count = this.int();
        const collections: OsdbData["collections"] = [];

        for (let collection_index = 0; collection_index < count; collection_index++) {
            const name = this.plain_string();
            const online_id = version_code >= 7 ? this.int() : 0;
            const beatmaps_count = this.int();
            const beatmaps: OsdbData["collections"][number]["beatmaps"] = [];

            for (let beatmap_index = 0; beatmap_index < beatmaps_count; beatmap_index++) {
                const difficulty_id = this.int();
                const beatmapset_id = version_code >= 2 ? this.int() : -1;

                const artist = is_minimal ? "" : this.plain_string();
                const title = is_minimal ? "" : this.plain_string();
                const difficulty = is_minimal ? "" : this.plain_string();
                const checksum = this.plain_string();

                const user_comment = version_code >= 4 ? this.plain_string() : "";
                const mode = version_code >= 8 || (version_code >= 5 && !is_minimal) ? this.byte() : 0;
                const difficulty_rating = version_code >= 8 || (version_code >= 6 && !is_minimal) ? this.double() : 0;

                beatmaps.push({
                    difficulty_id,
                    beatmapset_id,
                    artist,
                    title,
                    difficulty,
                    checksum,
                    user_comment,
                    mode,
                    difficulty_rating
                });
            }

            const hash_only_beatmaps: string[] = [];
            if (version_code >= 3) {
                const hashes_count = this.int();
                for (let hash_index = 0; hash_index < hashes_count; hash_index++) {
                    hash_only_beatmaps.push(this.plain_string());
                }
            }

            collections.push({
                name,
                online_id,
                beatmaps,
                hash_only_beatmaps
            });
        }

        const footer = this.plain_string();
        if (footer != OSDB_FOOTER) {
            throw new Error("invalid osdb footer");
        }

        this.data = {
            version_string,
            save_data,
            last_editor,
            count,
            collections
        };

        return this;
    };

    get = (): OsdbData => {
        return {
            version_string: this.data.version_string,
            save_data: this.data.save_data,
            last_editor: this.data.last_editor,
            count: this.data.count,
            collections: this.data.collections.map((collection) => ({
                name: collection.name,
                online_id: collection.online_id,
                beatmaps: collection.beatmaps.map((beatmap) => ({ ...beatmap })),
                hash_only_beatmaps: [...collection.hash_only_beatmaps]
            }))
        };
    };

    update = (patch: Partial<OsdbData>): this => {
        this.data = {
            ...this.data,
            ...patch,
            collections: patch.collections
                ? patch.collections.map((collection) => ({
                      name: collection.name,
                      online_id: collection.online_id,
                      beatmaps: collection.beatmaps.map((beatmap) => ({ ...beatmap })),
                      hash_only_beatmaps: [...collection.hash_only_beatmaps]
                  }))
                : this.data.collections
        };

        this.data.count = this.data.collections.length;
        return this;
    };

    write = async (): Promise<void> => {
        if (!this.location) {
            throw new Error("osdb parser has no target location");
        }

        const version_string = this.data.version_string || "o!dm8min";
        const version_code = map_version_to_code(version_string);
        const is_minimal = is_minimal_version(version_string);

        const payload: Buffer[] = [];
        payload.push(this.write_long(this.data.save_data));
        payload.push(this.write_plain_string(this.data.last_editor));
        payload.push(this.write_int(this.data.collections.length));

        for (const collection of this.data.collections) {
            payload.push(this.write_plain_string(collection.name));

            if (version_code >= 7) {
                payload.push(this.write_int(collection.online_id || 0));
            }

            payload.push(this.write_int(collection.beatmaps.length));

            for (const beatmap of collection.beatmaps) {
                payload.push(this.write_int(beatmap.difficulty_id || 0));

                if (version_code >= 2) {
                    payload.push(this.write_int(beatmap.beatmapset_id || 0));
                }

                if (!is_minimal) {
                    payload.push(this.write_plain_string(beatmap.artist || ""));
                    payload.push(this.write_plain_string(beatmap.title || ""));
                    payload.push(this.write_plain_string(beatmap.difficulty || ""));
                }

                payload.push(this.write_plain_string(beatmap.checksum || ""));

                if (version_code >= 4) {
                    payload.push(this.write_plain_string(beatmap.user_comment || ""));
                }

                if (version_code >= 8 || (version_code >= 5 && !is_minimal)) {
                    payload.push(this.write_byte(beatmap.mode || 0));
                }

                if (version_code >= 8 || (version_code >= 6 && !is_minimal)) {
                    payload.push(this.write_double(beatmap.difficulty_rating || 0));
                }
            }

            if (version_code >= 3) {
                payload.push(this.write_int(collection.hash_only_beatmaps.length));

                for (const hash of collection.hash_only_beatmaps) {
                    payload.push(this.write_plain_string(hash || ""));
                }
            }
        }

        payload.push(this.write_plain_string(OSDB_FOOTER));

        const version_buffer = this.write_plain_string(version_string);
        const payload_buffer = this.join_buffer(payload);

        if (uses_gzip_payload(version_code)) {
            const inner = this.join_buffer([version_buffer, payload_buffer]);
            const compressed = zlib.gzipSync(inner);
            const output = this.join_buffer([version_buffer, Buffer.from(compressed)]);
            fs.writeFileSync(this.location, output);
            return;
        }

        const output = this.join_buffer([version_buffer, payload_buffer]);
        fs.writeFileSync(this.location, output);
    };

    free = (): void => {
        this.location = "";
        this.set_buffer(Buffer.alloc(0));
        this.data = {
            version_string: "o!dm8min",
            save_data: 0n,
            last_editor: "",
            count: 0,
            collections: []
        };
    };
}

File: src/parser/osdb/osdb.cpp
#include <chrono>
#include <cstring>
#include <unordered_map>

#include "../../utils/binary.hpp"
#include "../../utils/gzip.hpp"
#include "osdb.hpp"

static bool ends_with(const std::string& value, const char* suffix) {
    const size_t suffix_len = std::strlen(suffix);

    if (value.size() < suffix_len) {
        return false;
    }

    return value.compare(value.size() - suffix_len, suffix_len, suffix) == 0;
}

static int osdb_version_to_code(const std::string& version) {
    static const std::unordered_map<std::string, int> version_map = {
        {"o!dm", 1},  {"o!dm2", 2}, {"o!dm3", 3}, {"o!dm4", 4},       {"o!dm5", 5},
        {"o!dm6", 6}, {"o!dm7", 7}, {"o!dm8", 8}, {"o!dm7min", 1007}, {"o!dm8min", 1008},
    };

    const auto it = version_map.find(version);
    return it != version_map.end() ? it->second : 0;
}

bool osdb_parser::parse(std::string location) {
    if (data == nullptr) {
        return false;
    }

    std::string target_location = std::move(location);
    std::vector<uint8_t> buffer;

    if (!binary::read_file_buffer(target_location, buffer)) {
        return false;
    }

    try {
        binary::BinaryCursor cursor;
        binary::set_cursor(cursor, buffer);

        const std::string version_string = binary::read_string2(cursor);
        const int version = osdb_version_to_code(version_string);

        if (version == 0) {
            return false;
        }

        const bool is_minimal = ends_with(version_string, "min");
        OsdbData temp;

        temp.version_string = version_string;
        std::vector<uint8_t> decompressed;

        if (version >= 7) {
            std::vector<uint8_t> compressed(buffer.begin() + static_cast<std::ptrdiff_t>(cursor.offset), buffer.end());

            if (!binary::gzip_decompress(compressed, decompressed)) {
                return false;
            }

            binary::set_cursor(cursor, decompressed);
            binary::read_string2(cursor);
        }

        temp.save_data = binary::read_i64(cursor);
        temp.last_editor = binary::read_string2(cursor);
        temp.count = binary::read_i32(cursor);

        if (temp.count < 0) {
            return false;
        }

        temp.collections.clear();
        temp.collections.reserve(static_cast<size_t>(temp.count));

        for (int i = 0; i < temp.count; i++) {
            OsdbCollection collection;
            collection.name = binary::read_string2(cursor);

            if (version >= 7) {
                collection.online_id = binary::read_i32(cursor);
            } else {
                collection.online_id = 0;
            }

            const int beatmaps_count = binary::read_i32(cursor);

            if (beatmaps_count < 0) {
                return false;
            }

            collection.beatmaps.clear();
            collection.beatmaps.reserve(static_cast<size_t>(beatmaps_count));

            for (int j = 0; j < beatmaps_count; j++) {
                OsdbBeatmap beatmap;

                beatmap.difficulty_id = binary::read_i32(cursor);
                beatmap.beatmapset_id = version >= 2 ? binary::read_i32(cursor) : -1;

                if (!is_minimal) {
                    beatmap.artist = binary::read_string2(cursor);
                    beatmap.title = binary::read_string2(cursor);
                    beatmap.difficulty = binary::read_string2(cursor);
                }

                beatmap.checksum = binary::read_string2(cursor);

                if (version >= 4) {
                    beatmap.user_comment = binary::read_string2(cursor);
                }

                if (version >= 8 || (version >= 5 && !is_minimal)) {
                    beatmap.mode = binary::read_u8(cursor);
                }

                if (version >= 8 || (version >= 6 && !is_minimal)) {
                    beatmap.difficulty_rating = binary::read_f64(cursor);
                }

                collection.beatmaps.push_back(std::move(beatmap));
            }

            if (version >= 3) {
                const int hash_count = binary::read_i32(cursor);
                collection.hash_only_beatmaps.clear();

                if (hash_count < 0) {
                    return false;
                }

                collection.hash_only_beatmaps.reserve(static_cast<size_t>(hash_count));

                for (int j = 0; j < hash_count; j++) {
                    collection.hash_only_beatmaps.push_back(binary::read_string2(cursor));
                }
            }

            temp.collections.push_back(std::move(collection));
        }

        const std::string footer = binary::read_string2(cursor);

        if (footer != "By Piotrekol") {
            return false;
        }

        *data = std::move(temp);
        return true;
    } catch (const std::exception& e) {
        return false;
    }
}

bool osdb_parser::write() {
    if (data == nullptr || location.empty()) {
        return false;
    }

    const std::string version_string = data->version_string.empty() ? "o!dm8min" : data->version_string;
    const int version = osdb_version_to_code(version_string);

    if (version == 0) {
        return false;
    }

    const bool is_minimal = ends_with(version_string, "min");

    std::vector<uint8_t> content;
    if (version >= 7) {
        binary::write_string2(content, version_string);
    }

    const int64_t save_time = data->save_data != 0
                                  ? data->save_data
                                  : static_cast<int64_t>(std::chrono::duration_cast<std::chrono::milliseconds>(
                                                             std::chrono::system_clock::now().time_since_epoch())
                                                             .count());

    data->count = static_cast<int>(data->collections.size());

    binary::write_i64(content, save_time);
    binary::write_string2(content, data->last_editor);
    binary::write_i32(content, data->count);

    for (const auto& collection : data->collections) {
        binary::write_string2(content, collection.name);

        if (version >= 7) {
            binary::write_i32(content, collection.online_id);
        }

        binary::write_i32(content, static_cast<int>(collection.beatmaps.size()));

        for (const auto& beatmap : collection.beatmaps) {
            binary::write_i32(content, beatmap.difficulty_id);

            if (version >= 2) {
                binary::write_i32(content, beatmap.beatmapset_id);
            }

            if (!is_minimal) {
                binary::write_string2(content, beatmap.artist);
                binary::write_string2(content, beatmap.title);
                binary::write_string2(content, beatmap.difficulty);
            }

            binary::write_string2(content, beatmap.checksum);

            if (version >= 4) {
                binary::write_string2(content, beatmap.user_comment);
            }

            if (version >= 8 || (version >= 5 && !is_minimal)) {
                binary::write_u8(content, static_cast<uint8_t>(beatmap.mode));
            }

            if (version >= 8 || (version >= 6 && !is_minimal)) {
                binary::write_f64(content, beatmap.difficulty_rating);
            }
        }

        if (version >= 3) {
            binary::write_i32(content, static_cast<int>(collection.hash_only_beatmaps.size()));
            for (const auto& hash : collection.hash_only_beatmaps) {
                binary::write_string2(content, hash);
            }
        }
    }

    binary::write_string2(content, "By Piotrekol");

    std::vector<uint8_t> buffer;
    binary::write_string2(buffer, version_string);

    if (version >= 7) {
        std::vector<uint8_t> compressed;

        if (!binary::gzip_compress(content, compressed)) {
            return false;
        }

        buffer.insert(buffer.end(), compressed.begin(), compressed.end());
    } else {
        buffer.insert(buffer.end(), content.begin(), content.end());
    }

    if (!binary::write_file_buffer(location, buffer)) {
        return false;
    }

    return true;
}
